### ISSUE TITLE
[2.0.1+] FontManagment.get_FontFiles: Cannot access a disposed object

### ISSUE THREAD

**Issue by sisve** [baskren/Forms9Patch#39]
The commit https://github.com/baskren/Forms9Patch/commit/2f3dde31e6c1c0027ba17967c95d79244bd5ad32#diff-09e2de900687e27e584f60d857e325e2R251 started disposing of cachedFontDir within the loop, which means that the second time the loop runs the invokation of `cachedFontDir.Exists()` will fail with the below message. The fix seems to be to move the call to Dispose to below the if-statement.

This seems to affect 2.0.1 and later versions. I've tested up to 2.2.2. Downgrading to 2.0.0 (before the commit) solves the problem.

	Cannot access a disposed object.
	Object name: 'Java.IO.File'.

	  at Java.Interop.JniPeerMembers.AssertSelf (Java.Interop.IJavaPeerable self) [0x00029] in <e7e2d009b69d4e5f9a00e6ee600b8a8e>:0 
	  at Java.Interop.JniPeerMembers+JniInstanceMethods.InvokeVirtualBooleanMethod (System.String encodedMember, Java.Interop.IJavaPeerable self, Java.Interop.JniArgumentValue* parameters) [0x00000] in <e7e2d009b69d4e5f9a00e6ee600b8a8e>:0 
	  at Java.IO.File.Exists () [0x0000a] in <11f101b564894ca7af6c482ddc51c698>:0 
	  at Forms9Patch.Droid.FontManagment.get_FontFiles () [0x00051] in C:\Users\ben\Development\Xamarin\Forms9Patch\Forms9Patch\Forms9Patch.Droid\Models\FontManagment.cs:233 
	  at Forms9Patch.Droid.FontManagment.TrySystemFont (System.String fontFamily) [0x00000] in C:\Users\ben\Development\Xamarin\Forms9Patch\Forms9Patch\Forms9Patch.Droid\Models\FontManagment.cs:35 
	  at Forms9Patch.Droid.FontManagment.TypefaceForFontFamily (System.String fontFamilys, System.Reflection.Assembly assembly) [0x000ad] in C:\Users\ben\Development\Xamarin\Forms9Patch\Forms9Patch\Forms9Patch.Droid\Models\FontManagment.cs:137 
	  at Forms9Patch.Droid.LabelRenderer.UpdateFont (Forms9Patch.Droid.F9PTextView control) [0x0000d] in C:\Users\ben\Development\Xamarin\Forms9Patch\Forms9Patch\Forms9Patch.Droid\Renderers\LabelRenderer.cs:660 
	  at Forms9Patch.Droid.LabelRenderer.InitControl (Forms9Patch.Droid.F9PTextView control) [0x00003] in C:\Users\ben\Development\Xamarin\Forms9Patch\Forms9Patch\Forms9Patch.Droid\Renderers\LabelRenderer.cs:508 
	  at Forms9Patch.Droid.LabelRenderer.OnElementChanged (Xamarin.Forms.Platform.Android.ElementChangedEventArgs`1[TElement] e) [0x00172] in C:\Users\ben\Development\Xamarin\Forms9Patch\Forms9Patch\Forms9Patch.Droid\Renderers\LabelRenderer.cs:470 
	  at Xamarin.Forms.Platform.Android.VisualElementRenderer`1[TElement].SetElement (TElement element) [0x000b8] in D:\a\1\s\Xamarin.Forms.Platform.Android\VisualElementRenderer.cs:231 
	  at Xamarin.Forms.Platform.Android.VisualElementRenderer`1[TElement].Xamarin.Forms.Platform.Android.IVisualElementRenderer.SetElement (Xamarin.Forms.VisualElement element) [0x00027] in D:\a\1\s\Xamarin.Forms.Platform.Android\VisualElementRenderer.cs:128 
	  at Xamarin.Forms.Platform.Android.Platform.CreateRenderer (Xamarin.Forms.VisualElement element, Android.Content.Context context) [0x0001f] in D:\a\1\s\Xamarin.Forms.Platform.Android\Platform.cs:344 
	  at Xamarin.Forms.Platform.Android.VisualElementPackager.AddChild (Xamarin.Forms.VisualElement view, Xamarin.Forms.Platform.Android.IVisualElementRenderer oldRenderer, Xamarin.Forms.Platform.Android.RendererPool pool, System.Boolean sameChildren) [0x000af] in D:\a\1\s\Xamarin.Forms.Platform.Android\VisualElementPackager.cs:138 
	  at Xamarin.Forms.Platform.Android.VisualElementPackager.SetElement (Xamarin.Forms.VisualElement oldElement, Xamarin.Forms.VisualElement newElement) [0x00139] in D:\a\1\s\Xamarin.Forms.Platform.Android\VisualElementPackager.cs:334 
	  at Xamarin.Forms.Platform.Android.VisualElementPackager.Load () [0x00000] in D:\a\1\s\Xamarin.Forms.Platform.Android\VisualElementPackager.cs:110 
	  at Xamarin.Forms.Platform.Android.VisualElementRenderer`1[TElement].SetPackager (Xamarin.Forms.Platform.Android.VisualElementPackager packager) [0x00007] in D:\a\1\s\Xamarin.Forms.Platform.Android\VisualElementRenderer.cs:440 
	  at Xamarin.Forms.Platform.Android.VisualElementRenderer`1[TElement].SetElement (TElement element) [0x000d5] in D:\a\1\s\Xamarin.Forms.Platform.Android\VisualElementRenderer.cs:234 
	  at Xamarin.Forms.Platform.Android.VisualElementRenderer`1[TElement].Xamarin.Forms.Platform.Android.IVisualElementRenderer.SetElement (Xamarin.Forms.VisualElement element) [0x00027] in D:\a\1\s\Xamarin.Forms.Platform.Android\VisualElementRenderer.cs:128 
	  at Xamarin.Forms.Platform.Android.Platform.CreateRenderer (Xamarin.Forms.VisualElement element, Android.Content.Context context) [0x0001f] in D:\a\1\s\Xamarin.Forms.Platform.Android\Platform.cs:344 
	  at Xamarin.Forms.Platform.Android.VisualElementPackager.AddChild (Xamarin.Forms.VisualElement view, Xamarin.Forms.Platform.Android.IVisualElementRenderer oldRenderer, Xamarin.Forms.Platform.Android.RendererPool pool, System.Boolean sameChildren) [0x000af] in D:\a\1\s\Xamarin.Forms.Platform.Android\VisualElementPackager.cs:138 
	  at Xamarin.Forms.Platform.Android.VisualElementPackager.SetElement (Xamarin.Forms.VisualElement oldElement, Xamarin.Forms.VisualElement newElement) [0x00139] in D:\a\1\s\Xamarin.Forms.Platform.Android\VisualElementPackager.cs:334 
	  at Xamarin.Forms.Platform.Android.VisualElementPackager.Load () [0x00000] in D:\a\1\s\Xamarin.Forms.Platform.Android\VisualElementPackager.cs:110 
	  at Xamarin.Forms.Platform.Android.VisualElementRenderer`1[TElement].SetPackager (Xamarin.Forms.Platform.Android.VisualElementPackager packager) [0x00007] in D:\a\1\s\Xamarin.Forms.Platform.Android\VisualElementRenderer.cs:440 
	  at Xamarin.Forms.Platform.Android.VisualElementRenderer`1[TElement].SetElement (TElement element) [0x000d5] in D:\a\1\s\Xamarin.Forms.Platform.Android\VisualElementRenderer.cs:234 
	  at Xamarin.Forms.Platform.Android.VisualElementRenderer`1[TElement].Xamarin.Forms.Platform.Android.IVisualElementRenderer.SetElement (Xamarin.Forms.VisualElement element) [0x00027] in D:\a\1\s\Xamarin.Forms.Platform.Android\VisualElementRenderer.cs:128 
	  at Xamarin.Forms.Platform.Android.Platform.CreateRenderer (Xamarin.Forms.VisualElement element, Android.Content.Context context) [0x0001f] in D:\a\1\s\Xamarin.Forms.Platform.Android\Platform.cs:344 
	  at Xamarin.Forms.Platform.Android.VisualElementPackager.AddChild (Xamarin.Forms.VisualElement view, Xamarin.Forms.Platform.Android.IVisualElementRenderer oldRenderer, Xamarin.Forms.Platform.Android.RendererPool pool, System.Boolean sameChildren) [0x000af] in D:\a\1\s\Xamarin.Forms.Platform.Android\VisualElementPackager.cs:138 
	  at Xamarin.Forms.Platform.Android.VisualElementPackager.OnChildAdded (System.Object sender, Xamarin.Forms.ElementEventArgs e) [0x0000f] in D:\a\1\s\Xamarin.Forms.Platform.Android\VisualElementPackager.cs:203 
	  at Xamarin.Forms.Element.OnChildAdded (Xamarin.Forms.Element child) [0x0000f] in D:\a\1\s\Xamarin.Forms.Core\Element.cs:326 
	  at Xamarin.Forms.VisualElement.OnChildAdded (Xamarin.Forms.Element child) [0x00000] in D:\a\1\s\Xamarin.Forms.Core\VisualElement.cs:742 
	  at Xamarin.Forms.Layout`1[T].OnChildAdded (Xamarin.Forms.Element child) [0x00000] in D:\a\1\s\Xamarin.Forms.Core\Layout.cs:32 
	  at Xamarin.Forms.Layout.OnInternalAdded (Xamarin.Forms.View view) [0x0001d] in D:\a\1\s\Xamarin.Forms.Core\Layout.cs:462 
	  at Xamarin.Forms.Layout.InternalChildrenOnCollectionChanged (System.Object sender, System.Collections.Specialized.NotifyCollectionChangedEventArgs e) [0x00078] in D:\a\1\s\Xamarin.Forms.Core\Layout.cs:452 
	  at (wrapper delegate-invoke) <Module>.invoke_void_object_NotifyCollectionChangedEventArgs(object,System.Collections.Specialized.NotifyCollectionChangedEventArgs)
	  at System.Collections.ObjectModel.ObservableCollection`1[T].OnCollectionChanged (System.Collections.Specialized.NotifyCollectionChangedEventArgs e) [0x00018] in /Users/builder/jenkins/workspace/archive-mono/2019-08/android/release/external/corefx/src/System.ObjectModel/src/System/Collections/ObjectModel/ObservableCollection.cs:263 
	  at System.Collections.ObjectModel.ObservableCollection`1[T].OnCollectionChanged (System.Collections.Specialized.NotifyCollectionChangedAction action, System.Object item, System.Int32 index) [0x00000] in /Users/builder/jenkins/workspace/archive-mono/2019-08/android/release/external/corefx/src/System.ObjectModel/src/System/Collections/ObjectModel/ObservableCollection.cs:338 
	  at System.Collections.ObjectModel.ObservableCollection`1[T].InsertItem (System.Int32 index, T item) [0x0001a] in /Users/builder/jenkins/workspace/archive-mono/2019-08/android/release/external/corefx/src/System.ObjectModel/src/System/Collections/ObjectModel/ObservableCollection.cs:196 
	  at System.Collections.ObjectModel.Collection`1[T].Add (T item) [0x00020] in /Users/builder/jenkins/workspace/archive-mono/2019-08/android/release/external/corefx/src/Common/src/CoreLib/System/Collections/ObjectModel/Collection.cs:71 
	  at Xamarin.Forms.ObservableWrapper`2[TTrack,TRestrict].Add (TRestrict item) [0x0004b] in D:\a\1\s\Xamarin.Forms.Core\ObservableWrapper.cs:35

**Comment by baskren** [comment]
@sisve - Thank you for the great analysis and stack trace.  Since you did not include a demo app, I cannot be sure I've really addressed the bug.  With that in mind, please evaluate the [v2.2.3-pre1 pre-release package](https://www.nuget.org/packages/Forms9Patch/2.2.3-pre1) and let me know if it addresses the issue.  If it does or does not (either way), it would be very helpful if you could share a small demo project that replicates the above crash since I cannot.  I would like to get that into the test suite ASAP.

**Comment by sisve** [comment]
I can confirm that the changes in https://github.com/baskren/Forms9Patch/commit/71c8512b9af51e431bffb094f23b880e6a635c6d fixed the problem.

I'm not sure how you would like the demo project shared. I've zipped it into https://www.dropbox.com/s/8uxknuwzq4o1479/Forms9PatchRepo39.zip

The steps I did to build that demo project:
1. Create a new Xamarin.Forms Android project.
2. Update the Xamarin.Forms version to 4.4, and add a reference to Forms9Sharp 2.2.2 (which requires XF 4.4)
3. The loop in question iterates files in Assets/Fonts, so I added two empty files there; `Assets/Fonts/file1.txt` and `Assets/Fonts/file2.txt`. Both have Build Action = AndroidAsset. 
4. To trigger the font management logic; specify a FontFamily on a label. `<forms9Patch:Label Text="Hello world!" FontFamily="non_existent" />` is enough.

An ObjectDisposedException will be thrown when the Label is rendered.
